src/long.ts

```diff
@@ -1021,9 +1023,37 @@ export class Long {
     if (options && options.relaxed) return this.toNumber();
     return { $numberLong: this.toString() };
   }
-  static fromExtendedJSON(doc: { $numberLong: string }, options?: EJSONOptions): number | Long {
-    const result = Long.fromString(doc.$numberLong);
-    return options && options.relaxed ? result.toNumber() : result;
+  static fromExtendedJSON(
+    doc: { $numberLong: string },
+    options?: EJSONOptions
+  ): number | Long | bigint {
+    const defaults = { useBigInt64: false, relaxed: true };
+    const ejsonOptions = { ...defaults, ...options };
+    const decimalRegEx = /^(\+?0|(\+|-)?[1-9][0-9]*)$/;
+
+    if (doc.$numberLong.length > MAX_INT64_STRING_LENGTH) {
+      throw new BSONError('int64 string is too long');
+    }
+
+    if (!decimalRegEx.test(doc.$numberLong)) {
+      throw new BSONError('int64 string is not in decimal');
```

Comment:
Since we know the string is not huge at this point we can include it in the error message